scipy/optimize/optimize.py

```diff
@@ -618,6 +618,123 @@ def approx_fprime(xk, f, epsilon, *args):
     return _approx_fprime_helper(xk, f, epsilon, args=args)
 
 
+def approx_jacobian(f, x0, f0=None, eps=None, args=()):
+    """Compute finite difference approximation of the Jacobian matrix of a
+    vector-valued function.
+
+    If a function maps from R^n to R^m, its Jacobian matrix has shape m x n
+    with an element (i, j) being partial derivative of f[i] with respect to
+    x[j].
+
+    Parameters
+    ----------
+    f : callable
+        The function of which to estimate the Jacobian matrix. Must take float
+        or array_like as the first argument and return float or array_like.
+    x0 : array_like or float
+        The point at which to estimate the Jacobian matrix.
+    f0 : {None, float, array_like}, optional
+        If not None it is assumed to be equal to f(x0), in this case the
+        f(x0) is not called.
+    eps : float or None, optional
+        The fractional accuracy with which the function is computed. If None,
+        `eps` is assigned to the machine epsilon of dtype of f(x0).
+    args : tuple, optional
+        Additional arguments passed to `f`.
+
+    Returns
+    -------
+    J : array
+        Finite difference approximation of the Jacobian matrix. If n > 1 and
+        m > 1 then it's a 2d-array. If n = 1 or m = 1 then it's a 1d-array.
+        If n = 1 and m = 1 then it's a scalar.
+
+    Notes
+    -----
+    The derivatives are estimated by forward finite differences with the
+    step along i-th coordinate computed as
+    ``dx[i] = sqrt(eps) * max(1, abs(x[i]))``, which approximately minimizes
+    the sum of round-off and truncation errors. Refer to [1]_ for the
+    discussion.
+
+    References
+    ----------
+    .. [1] William H. Press et. al. "Numerical Recipes. The Art of Scientific
+           Computing. 3rd edition", p. 229
+    """
+    x0 = np.asarray(x0)
+    if f0 is None:
+        f0 = f(x0, *args)
+    f0 = np.asarray(f0)
+
+    if eps is None:
+        eps = np.finfo(f0.dtype).eps
```

Comment:
this is going to fail if `f0` is integer: consider `def f(x): return x` and `x0=1`. The correct incantation involves `np.inexact`:
https://github.com/scipy/scipy/pull/4869/files#diff-6350cc4c1e51348a315df19ccc4e5353R143